### PR TITLE
Replace bitcoind wallet with bdk wallet

### PR DESCRIPTION
**Pull request by da-kami**
Create and pair bitcoind and electrs testcontainers using the same docker network and volume.

Follow-up to be fixed on top: 

- [ ] Remove bitcoin-harness dependency and directly depend on https://github.com/thomaseizinger/rust-jsonrpc-client for the RPC client.

### CODE REVIEW

File: swap/src/config.rs
Comment (da-kami) [inline swap/src/config.rs:17]: Note: @rishflab tested using this URL in a test-project and it actually worked. Might want to remove the comment?

**Review by bonomat — COMMENTED** [review]
We should not forget that we are using the sync version right now but actually want to switch to async. 
When doing so, don't forget that the default features do not work with the async features: 
`features = [ "esplora", "async-interface"], default-features=false`

File: swap/src/bitcoin/wallet.rs
Comment (bonomat) [inline swap/src/bitcoin/wallet.rs:267]: Unless you really want to go all the way to electrum, the bdk-wallet itself has a function to get the network.  
```
self.inner.lock().await.network()
```

File: swap/tests/testutils/bitcoind.rs
Comment (bonomat) [inline swap/tests/testutils/bitcoind.rs:75]: Note: your bitcoind image will always mount a volume with this approach even if `with_volume` is not being called. 
It would be better to use `Option::None`, set it to `Some(xxx)` in `with_volume` and check in the function `volumes()` if it is set.

File: swap/tests/testutils/bitcoind.rs
Comment (bonomat) [inline swap/tests/testutils/bitcoind.rs:129]: To simplify the setup you could fall back to the default bitcoin path in the image which is in `/root/.bitcoin`. This would allow you to use the image without a mounted volume.

File: swap/tests/testutils/electrs.rs
Comment (bonomat) [inline swap/tests/testutils/electrs.rs:10]: Can probably be removed.

File: swap/tests/testutils/mod.rs
Comment (bonomat) [inline swap/tests/testutils/mod.rs:449]: Afaik our bitcoind image exports the default ports and you could rely on docker to provide you random ports. That way you would not run into random `address in use` issues. 
See https://github.com/testcontainers/testcontainers-rs/blob/fe51e06fb2c44b1d3e3fd627cedea3fd582487ef/src/core/container.rs#L81

File: swap/tests/testutils/bitcoind.rs
Comment (bonomat) [inline swap/tests/testutils/bitcoind.rs:125]: We should use constants for username and admin or let it be autogenerated as in our coblox-bitcoin-core image: 

https://github.com/testcontainers/testcontainers-rs/blob/fe51e06fb2c44b1d3e3fd627cedea3fd582487ef/src/images/coblox_bitcoincore.rs#L51-L96

Otherwise we run in the danger of random errors because things get changed in different places.

File: swap/tests/testutils/electrs.rs
Comment (bonomat) [inline swap/tests/testutils/electrs.rs:136]: Most of the defaults don't make sense as electrum-rs won't work then.
Afaik, electrum-rs has some smart defaults which would just connect to public bitcoin nodes.

File: swap/tests/testutils/mod.rs
Comment (bonomat) [inline swap/tests/testutils/mod.rs:462]: See comment on the image in regards to username and password.

**Review by bonomat — CHANGES_REQUESTED** [review]
Nice to see the test pass 👏👏.

I think there are heaps of minor things which have to be ironed out before this is can be merged.

File: swap/tests/testutils/mod.rs
Comment (thomaseizinger) [inline swap/tests/testutils/mod.rs:494]: Passing the `Cli` into this function is weird. You are only using it in the last line of the function which means this function might as well just "create" the `Image` instance and return it, something like:

```rust
pub async fn new_electrs_image(
    volume: String,
    bitcoind_container_name: String,
) -> electrs::Electrs {
    let bitcoind_rpc_addr = format!("{}:18443", bitcoind_container_name);
    
    electrs::Electrs::default()
        .with_mapped_port(Port {
            local: 3012,
            internal: 3002,
        })
        .with_mapped_port(Port {
            local: 60401,
            internal: 60401,
        })
        .with_volume(volume)
        .with_daemon_rpc_addr(bitcoind_rpc_addr)
        .with_tag("latest")
}
```

That would simplify the signature quite a bit without majorly impacting the call-site.

**Review by thomaseizinger — APPROVED** [review]
Left some feedback, nothing is immediately pressing I think. Just some food for thought :)

File: swap/tests/testutils/electrs.rs
Comment (rishflab) [reply swap/tests/testutils/electrs.rs:136]: Are suggesting that ElectrsArgs should be an empty struct? This image is only used by the tests in these projects. Shouldn't the defaults reflect this?

File: swap/src/bitcoin/wallet.rs
Comment (rishflab) [reply swap/src/bitcoin/wallet.rs:64]: Yeh this has been noted in the list of issues. There was mention in the bdk docs of moving sync into a seperate thread and calling it in a loop.

Comment (da-kami) [reply swap/src/bitcoin/wallet.rs:64]: I propose the following change, instead of doing this in the wallet's balance function, where it clearly does not belong and most likely will cause blocking trouble.I would suggest to move it to the caller as @thomaseizinger suggested.

For the tests I see this right before we assert on balances. For production we might want to think about it a bit more, but as a start it might be sufficient to sync before and after a swap.

File: swap/src/bitcoin/wallet.rs
Comment (rishflab) [reply swap/src/bitcoin/wallet.rs:243]: nope, txindex no longer needed

File: swap/src/bitcoin/wallet.rs
Comment (rishflab) [reply swap/src/bitcoin/wallet.rs:43]: Need to implement conversion to anyhow::Error. Putting a todo: to do it later.

File: swap/tests/testutils/bitcoind.rs
Comment (rishflab) [reply swap/tests/testutils/bitcoind.rs:125]: switched to constants

File: swap/src/bitcoin/wallet.rs
Comment (thomaseizinger) [reply swap/src/bitcoin/wallet.rs:43]: Are you saying sled's errors don't implement `std::error::Error`? Because anything implementing that can be converted to `anyhow::Error`.

File: swap/src/bitcoin/transactions.rs
Comment (rishflab) [reply swap/src/bitcoin/transactions.rs:84]: ill leave a todo

File: swap/tests/testutils/electrs.rs
Comment (rishflab) [reply swap/tests/testutils/electrs.rs:57]: ill add it to the follow up tasks

File: swap/src/bitcoin/transactions.rs
Comment (rishflab) [reply swap/src/bitcoin/transactions.rs:49]: I left a todo to review this

File: swap/src/bitcoin/wallet.rs
Comment (rishflab) [reply swap/src/bitcoin/wallet.rs:175]: This expression ```
   self.get_tx(txid)
            .await
            .with_context(|| format!("Could not get raw tx with id: {}", txid))?
``` still returns an option

File: swap/src/protocol/alice.rs
Comment (rishflab) [reply swap/src/protocol/alice.rs:218]: I am leaving this for now because it stems from sending a psbt instead of a transaction in message0.

File: swap/src/bitcoin/wallet.rs
Comment (rishflab) [reply swap/src/bitcoin/wallet.rs:44]: Sled's errors are fine. The bdk electrum client error cant be handled with ?. It doesn't seem to implement `std::error::Error`

File: swap/tests/testutils/mod.rs
Comment (rishflab) [reply swap/tests/testutils/mod.rs]: We let docker map the ports now. To do that we need to create the `Container` which requires the `&Cli`.

File: swap/tests/testutils/electrs.rs
Comment (rishflab) [reply swap/tests/testutils/electrs.rs]: Used constants where I can added a todo to fix this later

**Review by rishflab — APPROVED** [review]

File: swap/src/bitcoin/timelocks.rs
Comment (D4nte) [inline swap/src/bitcoin/timelocks.rs:59]: Please clarify why you would substract two absolute block heights?

Also, implementing `sub_checked` would be safer.

File: swap/src/bitcoin/transactions.rs
Comment (D4nte) [inline swap/src/bitcoin/transactions.rs:21]: Using PSBT internally is for the best. I don't see the clone as an issue. I'd recommend to remove this TODO.

File: swap/src/bitcoin/wallet.rs
Comment (D4nte) [inline swap/src/bitcoin/wallet.rs:27]: The intent was for Bob to still use a `bitcoind`. It seems we are using BDK for both Alice and Bob. Is this intentional?

File: swap/src/bitcoin/wallet.rs
Comment (D4nte) [inline swap/src/bitcoin/wallet.rs:42]:
```suggestion
        let client =
            Client::new(electrum_rpc_url.as_str()).map_err(|e| anyhow!("Failed to init electrum rpc client: {:?}", e))?;
```
In the mean time you do not need to use `expect`.

File: swap/src/bitcoin/wallet.rs
Comment (D4nte) [inline swap/src/bitcoin/wallet.rs:44]: is `datadir`:
- The `data-dir` of the CLI? If so, we should but BDK related items in a subfolder (e.g. `bitcoin_wallet`.
- A subfolder of the CLI's `data-dir`? If so, then maybe best to call it `wallet-dir` so it's clear.

File: swap/src/bitcoin/wallet.rs
Comment (D4nte) [inline swap/src/bitcoin/wallet.rs:46]: Why would that be needed?

File: swap/src/bitcoin/wallet.rs
Comment (D4nte) [inline swap/src/bitcoin/wallet.rs:122]: Move this value in a const so it's easier to remove all static fee values at once.

File: swap/src/bitcoin/wallet.rs
Comment (D4nte) [inline swap/src/bitcoin/wallet.rs:140]: This log message does not seem very useful.

File: swap/src/bitcoin/wallet.rs
Comment (D4nte) [inline swap/src/bitcoin/wallet.rs:182]: Just do it?

File: swap/src/bitcoin/wallet.rs
Comment (D4nte) [inline swap/src/bitcoin/wallet.rs:224]: Sharing this error between the traits implementation would make it more readable.

File: swap/src/protocol/alice.rs
Comment (D4nte) [reply swap/src/protocol/alice.rs:218]: > I am leaving this for now because it stems from sending a psbt instead of a transaction in message0.

I do not understand your justification. Boxing it is very easy.

File: swap/tests/happy_path.rs
Comment (D4nte) [inline swap/tests/happy_path.rs:5]:
```suggestion
use swap::protocol::{alice, bob};
use testutils::SlowCancelConfig;
```

File: swap/src/bitcoin/wallet.rs
Comment (thomaseizinger) [reply swap/src/bitcoin/wallet.rs:175]: Fair, if you have to use `format!` it is probably equally noisy.

File: swap/src/protocol/alice.rs
Comment (thomaseizinger) [reply swap/src/protocol/alice.rs:218]: To be fair, a PSBT is probably the nicer solution anyway because that is the format designed for collaboratively signing a transaction.

**Review by thomaseizinger — APPROVED** [review]
Let's do it!

File: swap/src/bitcoin/wallet.rs
Comment (da-kami) [reply swap/src/bitcoin/wallet.rs:27]: Clarification: We intend for Alice to use `bitcoind` - as she is the role within Nectar. 

Since Nectar was not introduced yet we opted for changing both parties and then potentially bring back `bitcoind` for Alice again. (we discussed that in the chat)
I am not sure this strategy is optimal, as it might block merging https://github.com/comit-network/xmr-btc-swap/pull/177

@thomaseizinger what do you think? Should we just bring back `bitcoind` and maintain both wallets (one for CLI = Bob, one for Nectar = Alice)?

File: swap/src/bitcoin/wallet.rs
Comment (rishflab) [reply swap/src/bitcoin/wallet.rs:182]: Doable but not that straightforward. join can error and this function is not meant to fail.

File: swap/src/bitcoin/timelocks.rs
Comment (rishflab) [reply swap/src/bitcoin/timelocks.rs:59]: Yep switched to checked-sub this morning. Haven't pushed it.

We are subtracting the transaction block height with the latest block height to get the number of confirmations. BDK wallet does not have a API for this.

File: swap/src/bitcoin/wallet.rs
Comment (rishflab) [reply swap/src/bitcoin/wallet.rs]: mapping the error and propagating it now

File: swap/src/bitcoin/wallet.rs
Comment (thomaseizinger) [reply swap/src/bitcoin/wallet.rs:27]: Like we discussed on Friday, I wouldn't bother with bitcoind. Makes deploying Alice easier.

**Comment by thomaseizinger** [comment]
I think 88dcfaf should be squashed into f830e30.

Also, ee7496d seems to be generally useful and we can probably get that merged before we get the whole BDK stuff in. If that patch applies cleanly onto current master, it might be worth opening a separate PR for that.

**Comment by rishflab** [comment]
> I think [88dcfaf](https://github.com/comit-network/xmr-btc-swap/commit/88dcfaf266f9df65192e475844f4fcd65b21f3fa) should be squashed into [f830e30](https://github.com/comit-network/xmr-btc-swap/commit/f830e309d7558f795ab024248eaf704a661c6a81).
> 
> Also, [ee7496d](https://github.com/comit-network/xmr-btc-swap/commit/ee7496d794c8baea056d056115f15da25f614a08) seems to be generally useful and we can probably get that merged before we get the whole BDK stuff in. If that patch applies cleanly onto current master, it might be worth opening a separate PR for that.

Can do the squash. I don't think the last commit or equivalent commit can be merged or will make sense on its own.

**Comment by thomaseizinger** [comment]
> > I think [88dcfaf](https://github.com/comit-network/xmr-btc-swap/commit/88dcfaf266f9df65192e475844f4fcd65b21f3fa) should be squashed into [f830e30](https://github.com/comit-network/xmr-btc-swap/commit/f830e309d7558f795ab024248eaf704a661c6a81).
> > Also, [ee7496d](https://github.com/comit-network/xmr-btc-swap/commit/ee7496d794c8baea056d056115f15da25f614a08) seems to be generally useful and we can probably get that merged before we get the whole BDK stuff in. If that patch applies cleanly onto current master, it might be worth opening a separate PR for that.
> 
> Can do the squash. I don't think the last commit or equivalent commit can be merged or will make sense on its own.

Thank you! Did you intend to have the 2nd commit message as the "primary" one? The current commit message is just the concatenation of several commits. Do you mind re-wording that to a nice summary of what this patch actually does?

**Comment by rishflab** [comment]
> > > I think [88dcfaf](https://github.com/comit-network/xmr-btc-swap/commit/88dcfaf266f9df65192e475844f4fcd65b21f3fa) should be squashed into [f830e30](https://github.com/comit-network/xmr-btc-swap/commit/f830e309d7558f795ab024248eaf704a661c6a81).
> > > Also, [ee7496d](https://github.com/comit-network/xmr-btc-swap/commit/ee7496d794c8baea056d056115f15da25f614a08) seems to be generally useful and we can probably get that merged before we get the whole BDK stuff in. If that patch applies cleanly onto current master, it might be worth opening a separate PR for that.
> > 
> > 
> > Can do the squash. I don't think the last commit or equivalent commit can be merged or will make sense on its own.
> 
> Thank you! Did you intend to have the 2nd commit message as the "primary" one? The current commit message is just the concatenation of several commits. Do you mind re-wording that to a nice summary of what this patch actually does?

crap my bad, will fix

**Comment by da-kami** [comment]
I ran into this error on Bob's (CLI) side when swapping on test/stagenet with the current branch:

```
2021-02-15T11:43:07.926089Z  INFO swap::protocol::bob: Start execution setup with 12D3KooWCdMKjesXMJz1SiZ7HgotrxuqhQJbP5sgBm2BwP1cqThi
2021-02-15T11:43:19.720454Z DEBUG swap::bitcoin::wallet: building tx lock
2021-02-15T11:43:19.720502Z DEBUG swap::bitcoin::wallet: syncing wallet
2021-02-15T11:43:19.923615Z DEBUG swap::bitcoin::wallet: tx lock built
2021-02-15T11:43:20.001058Z  INFO swap{id=ae5a5a6b-7f25-4532-b64a-dbc4497679f9}: swap::protocol::bob::swap: Current state: negotiated
2021-02-15T11:43:20.001193Z DEBUG swap{id=ae5a5a6b-7f25-4532-b64a-dbc4497679f9}: swap::protocol::bob::event_loop: Attempt to dial Alice
2021-02-15T11:43:20.002030Z DEBUG swap::protocol::bob::event_loop: Already connected to Alice: 12D3KooWCdMKjesXMJz1SiZ7HgotrxuqhQJbP5sgBm2BwP1cqThi
2021-02-15T11:43:20.004695Z DEBUG swap{id=ae5a5a6b-7f25-4532-b64a-dbc4497679f9}: swap::bitcoin::wallet: signing tx lock: db990a50a375a0460e39f3e4e71ac454ab675a4e57f5fce5423ee12488bbdd2c
2021-02-15T11:43:20.008502Z DEBUG swap{id=ae5a5a6b-7f25-4532-b64a-dbc4497679f9}: swap::bitcoin::wallet: signed tx lock: db990a50a375a0460e39f3e4e71ac454ab675a4e57f5fce5423ee12488bbdd2c
2021-02-15T11:43:20.008862Z  INFO swap{id=ae5a5a6b-7f25-4532-b64a-dbc4497679f9}: swap::protocol::bob::state: db990a50a375a0460e39f3e4e71ac454ab675a4e57f5fce5423ee12488bbdd2c
2021-02-15T11:43:20.009080Z DEBUG swap{id=ae5a5a6b-7f25-4532-b64a-dbc4497679f9}: swap::bitcoin::wallet: attempting to broadcast tx: db990a50a375a0460e39f3e4e71ac454ab675a4e57f5fce5423ee12488bbdd2c
2021-02-15T11:43:20.206863Z  INFO swap{id=ae5a5a6b-7f25-4532-b64a-dbc4497679f9}: swap::bitcoin::wallet: Bitcoin tx broadcasted! TXID = db990a50a375a0460e39f3e4e71ac454ab675a4e57f5fce5423ee12488bbdd2c
2021-02-15T11:43:20.217932Z  INFO swap{id=ae5a5a6b-7f25-4532-b64a-dbc4497679f9}: swap::protocol::bob::swap: Current state: btc is locked
thread 'main' panicked at 'transient errors to be retried: JsonDeserialisation(reqwest::Error { kind: Decode, source: Error("expected value", line: 1, column: 1) })', swap/src/bitcoin/wallet.rs:238:10
stack backtrace:
```

It seems there can be other errors than the one we got with `confirmed` being missing.

**Comment by da-kami** [comment]
Additionally I ran into the problem, that the automated swap backend gets stuck when watching for the Bitcoin transaction:

```
2021-02-15T12:37:57.658433Z DEBUG swap::protocol::alice::event_loop: Connection Established with 12D3KooWFiXr3QNrAMFpQHkPVFDFJog6ULtAmbtzykkRo265Jotm
2021-02-15T12:37:57.660929Z DEBUG swap::protocol::alice::quote_response: Received quote request from 12D3KooWFiXr3QNrAMFpQHkPVFDFJog6ULtAmbtzykkRo265Jotm
2021-02-15T12:37:57.661137Z  INFO swap::protocol::alice::behaviour: Sent quote response
2021-02-15T12:38:01.060079Z  INFO swap::protocol::alice::behaviour: Start execution setup with 12D3KooWFiXr3QNrAMFpQHkPVFDFJog6ULtAmbtzykkRo265Jotm
2021-02-15T12:38:09.348981Z  WARN swap::protocol::alice::event_loop: Could not send swap handle over channel: SendError(RemoteHandle { rx: Receiver { inner: Inner { complete: false, data: Lock { locked: false, data: UnsafeCell }, rx_task: Lock { locked: false, data: UnsafeCell }, tx_task: Lock { locked: false, data: UnsafeCell } } }, keep_running: false })
2021-02-15T12:38:09.349221Z  INFO swap{id=c0a58da1-f672-4ed2-971a-03c5f2444e6d}: swap::protocol::alice::swap: Current state:started
2021-02-15T12:38:09.349387Z DEBUG swap{id=c0a58da1-f672-4ed2-971a-03c5f2444e6d}: swap::bitcoin::wallet: watching for tx: 921df66407eb6e29852570efaf3407693a0a3693224a97e3b989e7ed03d35d39
```

Before starting to watch there is a warning printed. I was not able to investigate if that is related to the program hanging.

**Comment by thomaseizinger** [comment]
> Before starting to watch there is a warning printed. I was not able to investigate if that is related to the program hanging.

I assume sure that error happens because we drop the receiving end of the swap handle in production because we only need that in the tests. @D4nte?

Should probably remove the warning because it is confusing.

**Comment by da-kami** [comment]
> > Before starting to watch there is a warning printed. I was not able to investigate if that is related to the program hanging.
> 
> I assume sure that error happens because we drop the receiving end of the swap handle in production because we only need that in the tests. @D4nte?
> 
> Should probably remove the warning because it is confusing.

I talked to @D4nte about it and yes, this code piece is only concerning the tests and in production the handle is dropped. I downgraded the message to `debug` level, but would push that on a separate branch once I have a fix for Alice's watching problem.

**Pull request merged**
